### PR TITLE
feat: Add support for AL2023 `nodeadm` user data

### LINKED ISSUE #2948: Support AL2023
## Is your request related to a new offering from AWS?
Is this functionality available in the [AWS provider for Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs)? See [CHANGELOG.md](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md), too.
  - Yes ✅: please list the AWS provider version which introduced this functionality

## Is your request related to a problem? Please describe.
AL2023 has already been launched in test mode, how can it be adapted to this module? Considering the changed approach to bootstrap instances. https://github.com/awslabs/amazon-eks-ami/issues/1672

## Describe the solution you'd like.
Transition from AL2 to AL2023 for all types of instances

### PR DESCRIPTION
**Pull request by bryantbiggs**
## Description
- Add support for AL2023 `nodeadm` user data

## Motivation and Context
- Ref https://github.com/awslabs/amazon-eks-ami/issues/1678
- Resolves #2948 

## Breaking Changes
- No

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [x] I have executed `pre-commit run -a` on my pull request

## ToDo
- [x] Remove `pre` and `post` bootstrap fields from AL2023 templates - this isn't valid/warranted with MIME multi-part
- [x] Add example nodegroup for both EKS and self-managed nodegroup
- [ ] Validate GA API

### CODE REVIEW

**Comment by bryantbiggs** [comment]
this is now GA so flipping the switch to `ready` https://aws.amazon.com/about-aws/whats-new/2024/02/amazon-eks-support-linux-2023/

File: modules/_user_data/main.tf [hunk 1]
```diff
@@ -1,65 +1,49 @@
...
       enable_bootstrap_user_data = var.enable_bootstrap_user_data
+
       # Required to bootstrap node
       cluster_name        = var.cluster_name
       cluster_endpoint    = var.cluster_endpoint
       cluster_auth_base64 = var.cluster_auth_base64
+
       # Optional
-      cluster_service_ipv4_cidr = var.cluster_service_ipv4_cidr != null ? var.cluster_service_ipv4_cidr : ""
+      cluster_service_ipv4_cidr = var.cluster_service_ipv4_cidr != null && contains(["linux", "al2023"], var.platform) ? var.cluster_service_ipv4_cidr : ""
```
Comment (antonbabenko) [inline modules/_user_data/main.tf:22]: Is this necessary to have `contains()` here? Normally, we ask users to specify values (`var.cluster_service_ipv4_cidr`) if they want, and they are responsible for whether it makes sense to be used there or not.

**Review by antonbabenko — APPROVED** [review]
Looks pretty good. Small question only.

File: modules/_user_data/main.tf (hunk 1 above)
Comment (bryantbiggs) [reply modules/_user_data/main.tf]: after re-visiting this, its a redundant guard that is not necessary. The operating systems that do not support this, they don't have the variable in their respective templates so we can remove the entire `contains()` - removed and updated examples to ensure no-op https://github.com/terraform-aws-modules/terraform-aws-eks/pull/2942/commits/202b43fb70b78e12c51fd598b709cacd7cce3081

**Comment by antonbabenko** [comment]
This PR is included in version 20.5.0 :tada:

**Pull request merged**